### ISSUE TITLE
Execution order of ReLU and Max-Pooling

### ISSUE THREAD

**Issue by Hvass-Labs** [tensorflow/tensorflow#3180]
Hello Everyone,

I'm new to Deep Learning and TensorFlow. From studying tutorials / research papers / online lectures it appears that people always have the execution order: ReLU -> Pooling. But in case of e.g. 2x2 max-pooling it seems that we can save 75% of the ReLU operations by simply reversing the execution order to: Max-Pooling -> ReLU. This should calculate the exact same thing using only a quarter of the ReLU operations. This reversal of operations can be done in general for max-pooling and all non-decreasing activation functions (which I guess they all are?), but it won't work for average-pooling.

This is an optimization that TensorFlow could perform automatically when compiling the computation graph. I haven't quite figured out how to use TensorBoard yet, so I can't tell if this automatic reversal of ReLU and max-pooling is already being done in TensorFlow.

So I've done a few experiments instead, timing the optimization of a convolutional net on MNIST, but the results are inconclusive for the two execution orders. Perhaps this means that TensorFlow already does the reversal automatically, or it means that there's no consistent advantage because the saved ReLU operations are such a tiny fraction of the overall computational cost, or perhaps it takes much larger images than MNIST and much deeper convolutional networks for the performance difference to become apparent.

Any thoughts?

**Comment by TimZaman** [comment]
Reversing the order 'optimizing' auto**magically** is a bad idea imo, especially (relatively to other DNN frameworks) tensorflow being relatively low level. Moreover, computationally, it might be inconclusive because the ReLU is a relatively cheap operation compared to the heavy convolutions, it would be a very minor optimization in terms of computational cost.
In any case, I think you're right in thinking that MaxPoolin(ReLU(x)) = ReLU(MaxPoolin(x)). I think this isn't the standard because one super-layer can be seen as {conv, ReLu} while the MaxPooling is an intermediate step to prepare for the next super-layer.

Tensorboard is a no-brainer to set-up initially, just run the `tensorboard --logdir=/foo/bar` and make sure you log your summary writer in `/foo/bar`, see the docs for this. You will be amaze: https://www.tensorflow.org/versions/r0.9/how_tos/summaries_and_tensorboard/index.html

**Comment by Hvass-Labs** [comment]
Such code optimizations are common in e.g. C / C++ compilers. It has probably been 15 years since I've implemented a compiler, but as I recall, these kinds of optimizations are called Peephole Optimizations. TensorFlow is a kind of compiler which uses a computational graph, so it would make sense for it to do automatic code optimizations like this.

Regarding the potential time savings, I think more investigation is needed before we rule it out. Let's do a quick back-of-the-envelope calculation to begin with. This may seem a little confusing and I hope I got the numbers right as I'm still new to these things.

The convolution performs approximately O(input_width * input_height * input_channels * filter_width * filter_height * num_filters) operations for each input image. This results in a tensor having approximately input_width * input_height * num_filters elements, depending on the padding settings. Dividing the two numbers shows that there's O(filter_width * filter_height * input_channels) arithmetic operations being performed to calculate one element of the output tensor. If we assume the ReLU operations take the same time to execute as each of those convolutional operations, then we should expect to save approximately 0.75 * (1 / (filter_width * filter_height * input_channels)) of the overall computational costs, because the cost of one single ReLU operation is approximately (1 / (filter_width * filter_height * input_channels)) of the total computational cost of the convolutional layer, and we can save 75% of those operations simply by switching the order of the ReLU operation and the 2x2 Max-Pooling.

For example, having filter_width == filter_height == 5 and input_channels == 1 we would get 0.75 * 0.04 = 0.03, that is, approximately 3% of the overall computational cost of the convolutional layer would be saved from this simple reversal of ReLU and Max-Pooling. That's actually a quite nice saving for such a simple code optimization! However, if input_channels == 64 then the saving is only 0.75 * (1 / 1600) which is about 0.0005 or about 0.05% which is clearly insignificant.

But if more expensive functions are used instead of ReLU, e.g. something with floating-point division which is computationally expensive, then perhaps it would still make sense to do this optimization even if the number of input channels is high.

Another thing to consider is the number of layers in the network, because the number of input channels to a convolutional layer is low in the first layer (e.g. 1-channel for gray-scale images and 3-channels for RGB colours), and the number of channels becomes higher in later layers because of the higher number of filter-channels. So the first layer might actually provide a small but tangible saving to the overall computational cost of the network, simply by switching the order of ReLU and Max-Pooling, while the deeper layers may only provide a tiny and insignificant computational saving. But if this optimization was done transparently by the TensorFlow compiler, then any potential time-saving would be gratis to the user of TensorFlow.

Nevertheless, I find it curious that people in Deep Learning continue to have the ReLU -> Max-Pooling ordering and apparently not realizing that it wastes operations. It suggests that people in the field might have a slightly rigid way of thinking about these things, and perhaps there are more substantial improvements waiting to be discovered.

**Comment by TimZaman** [comment]
Good write-up.
> It suggests that people in the field might have a slightly rigid way of thinking about these things

This is probably due to the empirical basis of many findings. We can also question why we are using MaxPooling at all? I have recently seen many nets where MaxPooling is substituted just by an extra convolutional stride, without significant loss of performance (if any).

**Comment by vrv** [comment]
Seems like a nice option for a future compiler optimization -- as @TimZaman said, I suspect this doesn't show up as a huge performance win yet, but these types of optimizations are definitely on our mind.

Btw, I would argue that when the user can't tell the difference, we should automatically do the optimization (e.g., if the only consumer of ReLU is MaxPooling -- no manual fetching or summaries, then it is probably safe to reorder them since the observed output is the same).

**Comment by Hvass-Labs** [comment]
@vrv Thanks for commenting on the dev-status on Peephole Optimizations, I was curious if it was already included or not in TensorFlow.

I've now tried PrettyTensor which automatically adds the activation function to the layer, e.g. like this for adding ReLU to each layer:

    with pt.defaults_scope(activation_fn=tf.nn.relu):
        y_pred, loss = x_pretty.\
            conv2d(kernel=5, depth=16, name='layer_conv1').\
            max_pool(kernel=2, stride=2).\
            conv2d(kernel=5, depth=36, name='layer_conv2').\
            max_pool(kernel=2, stride=2).\
            flatten().\
            fully_connected(size=128, name='layer_fc1').\
            softmax_classifier(class_count=10, labels=y_true)

I think it would make most sense to do Peephole Optimizations inside the TensorFlow compiler, rather than have people doing them manually in e.g. PrettyTensor, TFLearn, Keras, user-code, etc. As discussed above, there might not be a huge gain by reversing the activation function and max-pooling, but perhaps you will start to discover more possible optimizations once you see the code from this angle. I would be curious to hear of any discoveries - perhaps you can post in this thread if you find something significant?

@TimZaman Regarding max-pooling vs. stride in the convolutional layer, I did notice this redundancy when I was watching online lectures, reading tutorials and papers, etc. It is not a mathematical equivalency as in the case of execution order for relu/max-pool, so the TensorFlow compiler should not replace one for the other. But it would make sense that max-pooling and stride were quite similar, because nearby pixels in images are usually quite similar. If you have some references for papers that study max-pooling vs. stride then I'd like to see them. Thanks.

**Comment by aselle** [comment]
This would be a great optimization pass to contribute when XLA (our compiler framework is open soruced) [see here](https://www.tensorflow.org/versions/master/resources/xla_prerelease.html)

**Comment by drpngx** [comment]
Now that XLA is open-source, this might be worth reviving. Feel free to open a new issue referencing this one if you're ready to go!

**Comment by scanyameres** [comment]
@Hvass-Labs Your own explanation about the insignificant cost performance is very nice and correct, yet of course it is interesting to wonder if it would make sense to reverse the layer order anyways. IMHO I would expect industry or embedded-level networks to be optimized in such a way when possible, as any "free" computation saving is beneficial.

However, I wanted to point out that some networks DO need the ReLU to be performed right after the convolution. In semantic segmentation, for example, recurrent CNNs use the "convolution output" (including the ReLU) to feed again later stages of the architecture where an upsampling is required, and the original activations are necessary in the original size, to recover the exact neuron which fired the activation. If the MaxPool was done before the ReLU, this detailed local information would be lost, and therefore it wouldnt be possible to recover an output as big (or almost) as the original image.

I know it is not a big deal, and of course this only involves some small subset of CNNs, but hopefully I satisfied a bit of your curiosity! :)

**Comment by Hvass-Labs** [comment]
@scanyameres That is interesting. I didn't know that.
